### PR TITLE
[FLINK-15149][table-common] Merge InputTypeStrategy and InputTypeValidator

### PR DESCRIPTION
**Pull request by twalthr**
## What is the purpose of the change

This merged the concepts of `InputTypeStrategy` for inferring input data types and enriching data types with conversion classes and `InputTypeValidator` for validating arguments. It enables implicit casting and simplifies the overall design.

It contains a lot of refactoring/clean up and hopefully finalizes the type inference story for FLIP-65.

## Brief change log

- Merging of InputTypeStrategy and InputTypeValidator into InputTypeStrategy
- More tests and support for implicit casts
- Refactorings for consistent naming of the concepts

## Verifying this change

This change added tests and can be verified as follows:
`InputTypeStrategiesTest`

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: yes
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no
  - If yes, how is the feature documented? JavaDocs

### CODE REVIEW

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/inference/strategies/OrArgumentTypeStrategy.java [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+		final LogicalType actualType = callContext.getArgumentDataTypes().get(argumentPos).getLogicalType();
+
+		Optional<DataType> closestDataType = Optional.empty();
+		for (ArgumentTypeStrategy strategy : argumentStrategies) {
+			final Optional<DataType> inferredDataType = strategy.inferArgumentType(
+				callContext,
+				argumentPos,
+				false);
+			// argument type does not match at all
+			if (!inferredDataType.isPresent()) {
```
Comment (dawidwys) [inline flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/inference/strategies/OrArgumentTypeStrategy.java]: Can't we just break out of the loop here? Do we have to iterate over the remaining strategies?

Okay, after reading it again I see that it is a bit smarter and prefers a strategy that does not require an implicit cast. I think this further complicates the logic of the strategy, which is not obvious already. Can we just say it infers the first strategy that can be applied?

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/inference/ArgumentTypeStrategy.java [hunk 2]
```diff
@@ -21,24 +21,30 @@
...
-public interface ArgumentTypeValidator {
+public interface ArgumentTypeStrategy {
 
 	/**
-	 * Main logic for validating a single input type. Returns {@code true} if the argument is valid for the
-	 * given call, {@code false} otherwise.
+	 * Main logic for inferring and validating an argument. Returns the data type that is valid for
+	 * the given call. If the returned type differs from {@link CallContext#getArgumentDataTypes()} at
+	 * {@code argumentPos}, a casting operation can be inserted. An empty result means that the given
+	 * input is invalid.
```
[the remark this answers is not in the document]
Comment (twalthr) [reply flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/inference/ArgumentTypeStrategy.java]: Yes, you are right. It must not be invalid but just "invalid" depending on the available information. I changed the JavaDoc.

Regarding the additional method, I would like to keep the interfaces clear but will add an additional JavaDoc for the return type of the method for clarification.

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/inference/InputTypeStrategies.java [hunk 3]
```diff
@@ -20,49 +20,158 @@
...
+	 */
+	public static ExplicitArgumentTypeStrategy explicit(DataType expectedDataType) {
+		return new ExplicitArgumentTypeStrategy(expectedDataType);
+	}
 
 	/**
-	 * Input type strategy that supplies a fixed {@link DataType} for each argument.
+	 * Strategy for a conjunction of multiple {@link ArgumentTypeStrategy}s into one like
+	 * {@code f(NUMERIC && LITERAL)}. The strategy aims to infer the first more specific type or (if this
+	 * is not possible) an identical type.
```
[the remark this answers is not in the document]
Comment (twalthr) [reply flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/inference/InputTypeStrategies.java]: We cannot ensure that all argument strategies return the same more specific type. Usually one strategy can do that.

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/inference/InputTypeStrategies.java [hunk 4]
```diff
@@ -20,49 +20,158 @@
...
-	 * Special input type strategy for enriching data types with an expected conversion class. This
-	 * is in particular useful when a data type has been created out of a {@link LogicalType} but
-	 * runtime hints are still missing.
+	 * Strategy for a disjunction of multiple {@link ArgumentTypeStrategy}s into one like
+	 * {@code f(NUMERIC || STRING)}. The strategy aims to infer an identical type or (if this
+	 * is not possible) the first more specific type.
 	 */
-	public static BridgingInputTypeStrategy bridging(List<BridgingSignature> bridgingSignatures) {
-		return new BridgingInputTypeStrategy(bridgingSignatures);
+	public static OrArgumentTypeStrategy or(ArgumentTypeStrategy... strategies) {
```
[the remark this answers is not in the document]
Comment (twalthr) [reply flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/inference/InputTypeStrategies.java]: I improved the documentation here.

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/inference/TypeInference.java [hunk 5]
```diff
@@ -30,7 +30,7 @@
 /**
  * Provides logic for the type inference of function calls. It includes:
  * <ul>
- *     <li>explicit input specification for (possibly named) arguments</li>
+ *     <li>explicit input specification for (possibly named and/or typed) arguments</li>
```
[the remark this answers is not in the document]
Comment (twalthr) [reply flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/inference/TypeInference.java]: No, you can have named arguments but without typed arguments. If you just specify named arguments, they will be used for shuffling but no implicit casts will be inserted.

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/inference/TypeInference.java [hunk 6]
```diff
@@ -194,11 +178,8 @@ public Builder outputTypeStrategy(TypeStrategy outputTypeStrategy) {
 		}
 
 		public TypeInference build() {
-			return new TypeInference(
-				argumentNames,
-				argumentTypes,
+			return new TypeInference(namedArguments, typedArguments,
```
[the remark this answers is not in the document]
Comment (twalthr) [reply flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/inference/TypeInference.java]: Yes, this is possible. They are just used for reshuffling.

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/inference/strategies/OrArgumentTypeStrategy.java (hunk 1 above)
Comment (twalthr) [reply flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/inference/strategies/OrArgumentTypeStrategy.java]: Exactly, it prevents unnecessary casts. I updated the JavaDocs to make it more obvious. I would not return the first strategy that can be applied. All common languages including Java search for a matching type before starting type widening.

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/inference/InputTypeStrategies.java [hunk 7]
```diff
@@ -149,26 +149,38 @@ public static InputTypeStrategy or(InputTypeStrategy... strategies) {
...
 		return new AndArgumentTypeStrategy(Arrays.asList(strategies));
 	}
 
 	/**
 	 * Strategy for a disjunction of multiple {@link ArgumentTypeStrategy}s into one like
-	 * {@code f(NUMERIC || STRING)}. The strategy aims to infer an identical type or (if this
-	 * is not possible) the first more specific type.
+	 * {@code f(NUMERIC || STRING)}.
+	 *
+	 * <p>Some {@link ArgumentTypeStrategy}s cannot contribute an inferred type that is different from
```
Comment (dawidwys) [inline flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/inference/InputTypeStrategies.java]: I think this is really helpful!

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/inference/TypeStrategies.java [hunk 8]
```diff
@@ -47,11 +47,11 @@ public static TypeStrategy explicit(DataType dataType) {
 	}
 
 	/**
-	 * Type strategy that maps an {@link InputTypeValidator} to a {@link TypeStrategy} if the validator
-	 * matches.
+	 * Type strategy that maps an {@link InputTypeStrategy} to a {@link TypeStrategy} if the input strategy
```
[the remark this answers is not in the document]
Comment (dawidwys) [reply flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/inference/TypeStrategies.java]: After an offline chat. This comment does not make any sense. We do not know which input strategy was used before.

**Pull request closed without merging**